packages/bierzo-wallet/src/sequences/extension.ts

```diff
@@ -0,0 +1,29 @@
+/* global chrome */
+import { ThunkAction } from 'redux-thunk';
```

Comment:
This is complex, supposedly redux is a library for making your life easier, not complex. Sequences are something inherited from wallet-demo, which, as the `sanes-chrome-extension` has demonstrated are not necessary.

Keep in mind: In redux we should store information needed for customizing views. In this case, as you have defined, I would create a **user** reducer, using your still keep:
```
const initState: ExtensionState = {
  connected: false,
  installed: false,
};
```

Then have two actions:
```
SET_PERSONA_CONNECTED_ACTION <-- if instead of persona you prefer to use user, 
                                                                        for facing web app approach that is fine with me
SET_EXTENSION_INSTALLED_ACTION
```
On the other hand, please do not wrap chrome communication inside thunk actions, due its nature it is better for having controlled complexity create a folder named `communication` where you should place all logic needed for interact with the extension.

Something like:
```
src/communication/status.ts

export async function getExtensionStatus(): Promise<ExtensionState> {
  return new Promise(resolve => {
     try {
        chrome.runtime........
     ....
  })
}
```

Then what I would do in the src/index

```
getExtensionStatus()
  .then(status => {
     // create store with initial the just obtained status data
     //  mount react tree as usual
     //  navigate to welcome
  })
```

And finally in welcome route:
```
    const connected = useSelector(state => state.extension.connected)
    const installed = useSelector(state => state.extension.installed)
 
    ....
```

Maybe this big refactor sounds a bit strange at first, but you will be happy once you have finished it, because, when you start testing, you can easily mock using jest the `communication` folder.

And you avoid much difficult code.

Happy coding! 💪

Reply:
Yes, I think about it. But you should control state of the extension during all requests. Because user can remove application or logout somehow etc. So you need to be ready to update state during each request, that's why I decided to use thunk action. It will allow you to make request, update state in case of any operation result. And everything in one place. Unfortunately it is no possible to check extension state on during startup, you should do this all the time.

Reply:
Also I used same approach 
> SET_PERSONA_CONNECTED_ACTION <-- if instead of persona you prefer to use user, 
>                                                                        for facing web app approach that is fine with me
> SET_EXTENSION_INSTALLED_ACTION
But code start looks ugly because I was required to update both states (`connected`, `installed`) simultaneously in different places depends of operation result and decided to use one action for this

Reply:
This is a good and fair point. 

Let's start doing it on start-up. Once this is clean and done, we can evaluate to do the check on every request. We could do something like:

```
src/routes/foo/index.tsx

const Foo = ({...}: Props): JSX.Element => {
     ...  

    const onWhateverRequestToExtentionInSubmit = async (...) => {
        const extensionStatus = await getExtensionStatus();

       const somethingIsBad = !extensionStatus.connected || !extensionStatus.installed
       if (somethingIsBad) {
           // do whatever, update react, and force to navigate to other place where react is updated
          // or just throw error
       }
       
      await talkToExtension(....)
    }
}
```

But first, let's code the startup solution.

Reply:
There is also another issue, we should assume by default that everything is ok with the extension. In case if extension is ready and logged in we can not receive any extension status until user will accept/reject incoming identity sharing request.
Let me describe simple scenario. 
1. Extension was not installed -> we will get exception during `chrome.runtime.sendMessage(extensionId....` step
2. User was not logged in or Persona was not created -> we can understand this during 
```
const parsedResponse = parseJsonRpcResponse2(response);
        if (isJsonRpcErrorResponse(parsedResponse)) {
....
```` 
step
3. Everything is ok -> we can not get this state until user will click on Accept in the extension. So again we at the moment where we should set extension state.

so approach 
```
getExtensionStatus()
  .then(status => {
     // create store with initial the just obtained status data
     //  mount react tree as usual
     //  navigate to welcome
  })
```
will not work. Because in case if everything is ok `then` will be called only after user will accept share identity request. I think this is possible to avoid somehow but with usage of ugly hacks with `timeout` inside `getExtensionStatus()` Promise.

Reply:
That is right, good comment. In this case, we need to give a second thought. Seeing this project is a conversion of the old wallet-demo, we need to create the login view first, same as the wallet demo.

So, first is load the reducer using default props:
```
const initState: ExtensionState = {
  connected: false,
  installed: false,
};
```

Then in src/index.tsx
```
1. Create react instance as usual
2. Navigate to login view (in the future we'll navigate automatically to account view)
```

Third in login view:
```
const Login = ({...}) => {
  const dispatch = useDispatch();
  const onLogin = () => {
      toast.show("Please check your extension")
      getExtensionStatus()
        .then(status => {
            if (everythingIsOk) {
              // update redux
              // travel to welcome view, which will be refactored into account view in further steps
           }
           toast.show("Error!!");
        });
  }
}
```

Then in the `Header` component you can do:
```
    const connected = useSelector(state => state.extension.connected)
    const installed = useSelector(state => state.extension.installed)
```

And base on the connected and installed values, you can customize it.

Reply:
By the way, create a ticket for investigating what happens if you visit another view while the user interacts with the extension. Once the request is handled what would happen with the response? Is going to be received here?

Reply:
Yes, it should be received there. Also calling getExtensionStatus will help us only in case of exception. Because in this method we are relaying on requestShareIdentities API call